File: web/src/lib/fixtures.ts
/// Your fixtures with your answer to each (`GET /events/mine`,
/// `backend/db/src/repos/events.rs` MyFixture).
///
/// The fixtures list and the availability calendar both read this, so they
/// cannot disagree about whether you are playing.

import { api } from "@/lib/api";
import { dayKey } from "@/lib/availability";

export type Answer = "going" | "maybe" | "not_going";

export type MyFixture = {
  event_id: string;
  title: string;
  sport: string;
  event_subtype: string;
  status: string;
  start_at: string;
  end_at: string;
  club_id: string;
  club_name: string;
  opponent_club_id: string | null;
  opponent_club_name: string | null;
  venue_name: string | null;
  match_id: string | null;
  my_answer: Answer | null;
  fee_amount_cents: number | null;
  ticket_price_cents: number | null;
};

export const ANSWERS: { value: Answer; label: string; said: string }[] = [
  { value: "going", label: "Available", said: "You're available" },
  { value: "maybe", label: "Maybe", said: "You said maybe" },
  { value: "not_going", label: "Can't play", said: "You can't play" },
];

export function saidLabel(answer: Answer | null): string {
  return ANSWERS.find((a) => a.value === answer)?.said ?? "Not answered yet";
}

export function myFixtures(from: Date, to: Date): Promise<MyFixture[]> {
  const q = new URLSearchParams({ from: from.toISOString(), to: to.toISOString() });
  return api<MyFixture[]>("GET", `/events/mine?${q}`);
}

export function answerFixture(eventId: string, status: Answer): Promise<unknown> {
  return api("POST", `/events/${eventId}/rsvp`, { status });
}

/// Fixtures by the local calendar day they start on, in order.
export function byDay(fixtures: MyFixture[]): Map<string, MyFixture[]> {
  const days = new Map<string, MyFixture[]>();
  for (const f of fixtures) {
    const key = dayKey(new Date(f.start_at));
    days.set(key, [...(days.get(key) ?? []), f]);
  }
  return days;
}

/// Two fixtures you have said yes to that are on at the same time.
export function clashes(fixtures: MyFixture[]): Set<string> {
  const yes = fixtures.filter((f) => f.my_answer === "going");
  const out = new Set<string>();
  for (const a of yes)
    for (const b of yes)
      if (a !== b && Date.parse(a.start_at) < Date.parse(b.end_at) && Date.parse(b.start_at) < Date.parse(a.end_at))
        out.add(a.event_id);
  return out;
}

export function timeOf(iso: string): string {
  return new Date(iso).toLocaleTimeString("en-GB", { hour: "2-digit", minute: "2-digit" });
}

export function dayTitle(key: string): string {
  const [y, m, d] = key.split("-").map(Number);
  const date = new Date(y, m - 1, d);
  const today = dayKey(new Date());
  const tomorrow = dayKey(new Date(Date.now() + 864e5));
  const long = date.toLocaleDateString("en-GB", { weekday: "long", day: "numeric", month: "long" });
  return key === today ? `Today · ${long}` : key === tomorrow ? `Tomorrow · ${long}` : long;
}

File: ios/Fishers/Models/Fixtures.swift
import Foundation

/// "Can you play?" — the three answers, saved as an RSVP.
enum FixtureAnswer: String, Codable, CaseIterable, Identifiable {
    case going
    case maybe
    case notGoing = "not_going"

    var id: String { rawValue }

    var label: String {
        switch self {
        case .going: return "Available"
        case .maybe: return "Maybe"
        case .notGoing: return "Can't play"
        }
    }

    /// What the card says back.
    var said: String {
        switch self {
        case .going: return "You're available"
        case .maybe: return "You said maybe"
        case .notGoing: return "You can't play"
        }
    }

    var systemImage: String {
        switch self {
        case .going: return "checkmark.circle.fill"
        case .maybe: return "questionmark.circle.fill"
        case .notGoing: return "xmark.circle.fill"
        }
    }

    var rsvp: RsvpStatus {
        switch self {
        case .going: return .going
        case .maybe: return .maybe
        case .notGoing: return .notGoing
        }
    }
}

/// One of your fixtures with your answer to it (`GET /events/mine`). The list
/// and the calendar both read this, so they cannot disagree about whether you
/// are playing.
struct MyFixture: Codable, Identifiable, Hashable {
    let eventId: UUID
    var title: String
    var sport: String
    var eventSubtype: String
    var status: String
    var startAt: Date
    var endAt: Date
    var clubId: UUID
    var clubName: String
    var opponentClubId: UUID?
    var opponentClubName: String?
    var venueName: String?
    var matchId: UUID?
    var myAnswer: FixtureAnswer?
    var feeAmountCents: Int?
    var ticketPriceCents: Int?

    var id: UUID { eventId }

    enum CodingKeys: String, CodingKey {
        case title, sport, status
        case eventId = "event_id"
        case eventSubtype = "event_subtype"
        case startAt = "start_at"
        case endAt = "end_at"
        case clubId = "club_id"
        case clubName = "club_name"
        case opponentClubId = "opponent_club_id"
        case opponentClubName = "opponent_club_name"
        case venueName = "venue_name"
        case matchId = "match_id"
        case myAnswer = "my_answer"
        case feeAmountCents = "fee_amount_cents"
        case ticketPriceCents = "ticket_price_cents"
    }

    var saidLabel: String { myAnswer?.said ?? "Not answered yet" }

    var systemImage: String {
        switch eventSubtype {
        case "league_match", "tournament": return "trophy"
        case "friendly": return "figure.cricket"
        case "nets": return "sportscourt"
        case "social": return "person.3"
        default: return "calendar"
        }
    }

    var isCancelled: Bool { status == "cancelled" }
    var isPostponed: Bool { status == "postponed" }
}

enum FixtureList {
    /// Fixtures by the local calendar day they start on, days in order.
    static func byDay(_ fixtures: [MyFixture], calendar: Calendar = .current) -> [(day: Date, fixtures: [MyFixture])] {
        let grouped = Dictionary(grouping: fixtures) { calendar.startOfDay(for: $0.startAt) }
        return grouped.keys.sorted().map { day in
            (day, grouped[day]!.sorted { $0.startAt < $1.startAt })
        }
    }

    /// Fixtures you have said yes to that overlap another you said yes to.
    static func clashes(_ fixtures: [MyFixture]) -> Set<UUID> {
        let yes = fixtures.filter { $0.myAnswer == .going }
        var out = Set<UUID>()
        for a in yes {
            for b in yes where a.eventId != b.eventId && a.startAt < b.endAt && b.startAt < a.endAt {
                out.insert(a.eventId)
            }
        }
        return out
    }

    /// "Today · Sunday 14 September", "Tomorrow · …", or the long date.
    static func dayTitle(_ day: Date, now: Date = .now, calendar: Calendar = .current) -> String {
        let long = day.formatted(.dateTime.weekday(.wide).day().month(.wide))
        if calendar.isDate(day, inSameDayAs: now) { return "Today · \(long)" }
        if let tomorrow = calendar.date(byAdding: .day, value: 1, to: now), calendar.isDate(day, inSameDayAs: tomorrow) {
            return "Tomorrow · \(long)"
        }
        return long
    }

    /// Every date in `month` that falls on `weekday` (1 = Sunday … 7 = Saturday).
    static func dates(in month: Date, weekday: Int, calendar: Calendar = .current) -> [Date] {
        guard let range = calendar.range(of: .day, in: .month, for: month),
              let first = calendar.date(from: calendar.dateComponents([.year, .month], from: month))
        else { return [] }
        return range.compactMap { calendar.date(byAdding: .day, value: $0 - 1, to: first) }
            .filter { calendar.component(.weekday, from: $0) == weekday }
    }
}
